ptypy/custom/ePIE_multislice.py

```diff
@@ -0,0 +1,181 @@
+        super().engine_initialize()
+
+        # Create a list of objects and exit waves (one for each slice)
+        self._object = [None] * self.p.number_of_slices
+        self._probe = [None] * self.p.number_of_slices
+        self._exits = [None] * self.p.number_of_slices
+        for i in range(self.p.number_of_slices):
+            self._object[i] = self.ob.copy(self.ob.ID + "_o_" + str(i))
+            self._probe[i] = self.pr.copy(self.pr.ID + "_p_" + str(i))
+            self._exits[i] = self.pr.copy(self.pr.ID + "_e_" + str(i))
+
+        # ToDo:
+        #    - allow for non equal slice spacing
+        #    - allow for start_slice_update at a freely chosen iteration
+        #      for each slice separately
+
+        scan = list(self.ptycho.model.scans.values())[0]
+        geom = scan.geometries[0]
+        g = Param()
+        g.energy = geom.energy
+        g.distance = self.p.slice_thickness
+        g.psize = geom.resolution
+        g.shape = geom.shape
+        g.propagation = "nearfield"
+        G = geometry.Geo(owner=None, pars=g)
+        self.fw = G.propagator.fw
+        self.bw = G.propagator.bw
+
+    def engine_iterate(self, num=1):
+        """
+        Compute one iteration.
+        """
+        vieworder = list(self.di.views.keys())
+        vieworder.sort()
+        rng = np.random.default_rng()
+
+        for it in range(num):
+
+            error_dct = {}
+            rng.shuffle(vieworder)
+
+            for name in vieworder:
+                view = self.di.views[name]
+                if not view.active:
+                    continue
+
+                # Multislice update
+                error_dct[name] = self.multislice_update(view)
+
+            self.curiter += 1
+
+        return error_dct
+
+    def engine_finalize(self):
+        self.ob.fill(self._object[0])
+        for i in range(1, self.p.number_of_slices):
+            self.ob *= self._object[i]
+
+        # Save the slices
+        slices_info = Param()
+        slices_info.number_of_slices = self.p.number_of_slices
+        slices_info.slice_thickness = self.p.slice_thickness
+        slices_info.objects = {ob.ID: {ID: S._to_dict() for ID, S in ob.storages.items()}
+                               for ob in self._object}
+
+        header = {'description': 'multi-slices result details.'}
+
+        h5opt = io.h5options['UNSUPPORTED']
+        io.h5options['UNSUPPORTED'] = 'ignore'
+        logger.info(f'Saving to {self.p.fslices}')
+        io.h5write(self.p.fslices, header=header, content=slices_info)
+        io.h5options['UNSUPPORTED'] = h5opt
+
+        return super().engine_finalize()
+
+    def multislice_update(self, view):
+        """
+        Does multislice ePIE
+        based on https://doi.org/10.1364/JOSAA.29.001606
+        """
+
+        for name, pod in view.pods.items():
+            # Forward multislice, calculate exit waves
+            for i in range(self.p.number_of_slices-1):
+                # exit wave for this slice
```

Comment:
Thanks @kahntm for expanding this code to use modes! Maybe switch these loops around? Shouldn't make any difference to the calculations but its more consistent with the backwards update further below.

Reply:
Possible, but it would result in a 2nd loop over all pods in the view to update the the exit wave in the last slice and update the pods probe and object
![image](https://github.com/ptycho/ptypy/assets/26224125/69ed2bb8-0ac9-4ac3-9546-068e2a75b381)

Alternatively the loop would be extended one index further and and if i==self.p.number_of_slices-1 would need to be added within the inner loop

Reply:
I think having a second loop over the pods should be fine. I doubt it would make much of a difference performance wise...